### ISSUE TITLE
[ONNX] incorrect export of opset10  slice with dynamic_axis

### ISSUE THREAD

**Issue by dashesy** [pytorch/pytorch#28611]
## 🐛 Bug

When using `dynamic_axes` the `slice` operator (since opset10) always should use `shape` information instead of fixed constants. Currently fixed sizes are used and the only workaround is to use jit.

## To Reproduce

```python
import torch
import torch.nn as nn

@torch.jit.script
def center_slice_helper(x, h_offset, w_offset, h_end, w_end):
    return x[:, :, h_offset:h_end, w_offset:w_end]


class CenterCrop(nn.Module):
    def __init__(self, crop_size, use_jit=False):
        """Crop from the center of a 4d tensor
        Input shape can be dynamic
        :param crop_size: the center crop size
        :param use_jit: if should use the jit helpter
        """
        super(CenterCrop, self).__init__()
        self.crop_size = crop_size
        self.use_jit = use_jit
        self.register_buffer('crop_size_t', torch.tensor(crop_size))

    def extra_repr(self):
        """Extra information
        """
        return 'crop_size={}'.format(
            self.crop_size
        )

    def forward(self, x):
        """
        :type x: torch.Tensor
        """
        height, width = x.shape[2], x.shape[3]
        if not isinstance(height, torch.Tensor):
            height, width = torch.tensor(height).to(x.device), torch.tensor(width).to(x.device)
        h_offset = (height - self.crop_size_t) / 2
        w_offset = (width - self.crop_size_t) / 2
        h_end = h_offset + self.crop_size_t
        w_end = w_offset + self.crop_size_t
        if self.use_jit:
            return center_slice_helper(x, h_offset, w_offset, h_end, w_end)
        return x[:, :, h_offset:h_end, w_offset:w_end]

model = CenterCrop(224)
onnxfile = "/mnt/output/gr/crop.onnx"

targets = ["cropped"]
dynamic_axes = {'data': [2, 3]}
dummy_input = torch.randn(1, 3, 300, 256, device='cpu').byte()
torch.onnx.export(model, dummy_input, onnxfile,
                  verbose=True, input_names=['data'],
                  dynamic_axes=dynamic_axes,
                  output_names=targets,
                  opset_version=10)

```

Results in this fixed slice, even though dynamic_axes is used:

![FixedSliceOpset10](https://user-images.githubusercontent.com/873905/67514509-b4589980-f651-11e9-9df9-a81ea34ba374.JPG)

Now to work around this use jit:

```python
model = CenterCrop(224, use_jit=True)
onnxfile = "/mnt/output/gr/crop.onnx"

targets = ["cropped"]
dynamic_axes = {'data': [2, 3]}
dummy_input = torch.randn(1, 3, 300, 256, device='cpu').byte()
torch.onnx.export(model, dummy_input, onnxfile,
                  verbose=True, input_names=['data'],
                  dynamic_axes=dynamic_axes,
                  output_names=targets,
                  opset_version=10)

```

Which is correct:
![CorrectSliceOpset10](https://user-images.githubusercontent.com/873905/67514744-334dd200-f652-11e9-9cc3-9ce6183902a4.JPG)

## Expected behavior

Even without jit, `Slice` in opset10 should use shape information dynamically when `dynamic_axes` is used.

## Environment

PyTorch version: 1.3.0
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.14.0

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1070
Nvidia driver version: 418.40.04
cuDNN version: /usr/local/lib/libcudnn.so.5.1.10

Versions of relevant libraries:
[pip] numpy==1.17.2
[pip] torch==1.3.0
[pip] torchvision==0.4.1a0+d94043a
[conda] Could not collect

## Additional context

If instead we exported with opset9, still jit workaround works but uses `DynamicSlice`
```python
model = CenterCrop(224, use_jit=True)
onnxfile = "/mnt/output/gr/crop.onnx"

targets = ["cropped"]
dynamic_axes = {'data': [2, 3]}
dummy_input = torch.randn(1, 3, 300, 256, device='cpu').byte()
torch.onnx.export(model, dummy_input, onnxfile,
                  verbose=True, input_names=['data'],
                  dynamic_axes=dynamic_axes,
                  output_names=targets,
                  opset_version=9)
```

![DynamicSliceOpset9](https://user-images.githubusercontent.com/873905/67515115-e3bbd600-f652-11e9-949e-67fd593aac56.JPG)

`DynamicSlice` [is obsolete extension](https://github.com/microsoft/onnxruntime/issues/1326) now that [Slice](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Slice) since opset10 can be dynamic. But the workaround of jit makes this not very nice, specially that using jit has some [other issues](https://github.com/pytorch/pytorch/issues/28607).

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

**Comment by neginraoof** [comment]
@dashesy the fix for this was merged yesterday. I just tried your code snippet and I get the dynamic model exported. Are building the latest master?

**Comment by dashesy** [comment]
@neginraoof Thanks! I will try the nightly. So no need for `jit` workaround any more?

**Comment by neginraoof** [comment]
No that should not be needed.

**Comment by dashesy** [comment]
@neginraoof Yes this is fixed in the nightly! thanks.
Just one quirk, why do I still get this warning:

> TracerWarning: Converting a tensor to a Python index might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means tha t the trace might not generalize to other inputs!                                                                                                                                                                                                                                 return x[:, :, h_offset:h_end, w_offset:w_end]

While the exported model is actually correct:

![SliceOpset10](https://user-images.githubusercontent.com/873905/67527718-8896dd00-f66c-11e9-9b71-74d9d795ff33.JPG)


These warnings are quite useful if they are not false positives.

**Comment by neginraoof** [comment]
@dashesy This warning is at autograd level for slice indexing, not exporter directly. But we might be able to remove it now. I'll check that. I think we can close the issue.

**Comment by neginraoof** [comment]
cc @colesbury 
The issue for tracing slice index dynamically has been resolved here:
https://github.com/pytorch/pytorch/pull/26549
Do you think we can remove this warning?
https://github.com/pytorch/pytorch/blob/46f96d153864f6da798700d97eea8255d1e3ddf4/tools/autograd/templates/python_variable_methods.cpp#L312

**Comment by dashesy** [comment]
Is this change going to be included in 1.4? we would like to automate some scrips and rather avoid nightly build.

**Comment by neginraoof** [comment]
Yes, looks like this is included.
